src/arch/io/network.cc

```diff
@@ -621,6 +621,430 @@ void linux_tcp_conn_t::on_event(int /* events */) {
     event_watcher->stop_watching_for_errors();
 }
 
+/* tls_conn_wrapper_t wraps a TLS connection and safely releases it in
+exceptional cases if it is not reclaimed before execution leaves its scope. */
+class tls_conn_wrapper_t {
+public:
+    tls_conn_wrapper_t(SSL *_conn) {
+        conn = _conn;
+    }
+
+    ~tls_conn_wrapper_t() {
+        SSL_free(conn);
+    }
+
+    // Transfer ownership of the connection to the caller.
+    void reclaim() {
+        conn = NULL;
+    }
+
+private:
+    SSL *conn;
+};
+
+/* This is the client version of the constructor. The base class constructor
+will establish a TCP connection to the peer at the given host:port and then we
+wrap the tcp connection in TLS using the configuration in the given tls_ctx. */
+linux_secure_tcp_conn_t::linux_secure_tcp_conn_t(SSL_CTX *tls_ctx, const ip_address_t &host, int port, signal_t *interruptor, int local_port) THROWS_ONLY(connect_failed_exc_t, interrupted_exc_t) :
+        linux_tcp_conn_t(host, port, interruptor, local_port) {
+
+
+    conn = SSL_new(tls_ctx);
+
+    /* If establish_conn() throws an exception, the conn_wrapper deallocates
+    the connection when unwinding the stack. */
+    tls_conn_wrapper_t conn_wrapper(conn);
+
+    establish_conn(SSL_connect);
+
+    conn_wrapper.reclaim();
+}
+
+/* This is the server version of the constructor */
+linux_secure_tcp_conn_t::linux_secure_tcp_conn_t(SSL_CTX *tls_ctx, fd_t sock) :
+        linux_tcp_conn_t(sock) {
+
+    conn = SSL_new(tls_ctx);
+
+    try {
+        establish_conn(SSL_accept);
+    } catch (const tcp_conn_t::connect_failed_exc_t &err) {
+        // Ignore.
+        logNTC("linux_secure_tcp_conn_t server constructor exception: %s\n", err.info.c_str());
+    }
+}
+
+linux_secure_tcp_conn_t::~linux_secure_tcp_conn_t() THROWS_NOTHING {
+    assert_thread();
+
+    if (is_open()) shutdown();
+
+    SSL_free(conn);
+}
+
+/* Creates a new OpenSSL connection using the underlying fd and the given
+OpenSSL context. The given handshake function is used to perform the handshake
+for the client or server using SSL_connect or SSL_accept respectively. */
+void linux_secure_tcp_conn_t::establish_conn(int (*handshake)(SSL *) ) THROWS_ONLY(connect_failed_exc_t) {
+    // Add support for partial writes.
+    SSL_set_mode(conn, SSL_MODE_ENABLE_PARTIAL_WRITE);
+
+    if (NULL == conn) {
```

Comment:
I suggest moving this check into the `tls_conn_wrapper_t` constructor as mentioned above, though you can leave it this way if you prefer.